Remove grunt

noppanit opened the pull request:
As we don't need Grunt and Gulp, this PR attempts to replace Grunt with Gulp.

Related: https://github.com/HabitRPG/habitrpg/issues/5829

noppanit committed:
Deleted clean task from Gruntfile and add gulp clean to gulp

noppanit committed:
- Upgraded browserify to latest version
- Move browserify task from Grunt to Gulp

Note: Do not use gulp-browserify

noppanit committed:
Add gulp-minify-css

noppanit committed:
Add gulp-stylus

noppanit committed:
Add gulp copy task

crookedneighbor commented on tasks/gulp-build.js:
```diff
@@ -1,4 +1,12 @@
+let stylus = require('gulp-stylus');
+let nib = require('nib');
+
 require('gulp-grunt')(gulp);
```
Is this needed?

crookedneighbor commented on tasks/gulp-build.js:
```diff
@@ -10,7 +18,36 @@ gulp.task('build', () => {
+
+	b.transform('coffeeify').bundle().pipe(source('index.js')).pipe(rename('habitrpg-shared.js'))
+		.pipe(gulp.dest('common/dist/scripts/'))
+  //gulp.start('grunt-build:dev', done);
```
This should be removed.

noppanit committed:
Merged with upstream develop

crookedneighbor commented on tasks/gulp-clean.js:
```diff
@@ -0,0 +1,7 @@
+let rimraf = require('rimraf');
+
+gulp.task('clean', (cb) => {
+	rimraf('website/build', cb);
```
Our style is two spaces, not 4.

crookedneighbor commented on tasks/gulp-copy.js:
```diff
@@ -0,0 +1,11 @@
+import gulp from 'gulp';
+
+gulp.task('copy', (cb) => {
+	gulp.src('website/public/favicon.ico').pipe(gulp.dest('website/build'));
```
2 spaces, not 4.

crookedneighbor commented:
This is looking pretty good, just got some style notes for you.

Is there a reason you haven't deleted the other unused tasks for the gruntfile?

noppanit commented:
Sorry I forgot to clean up, I'll do it and push.

crookedneighbor commented:
What's the status on this?

noppanit commented:
It's going to be pushed soon

noppanit committed:
Clean up old code and change the tab to be 2 spaces

noppanit committed:
Merge branch 'develop' of https://github.com/HabitRPG/habitrpg into remove_grunt

noppanit committed:
Convert tab to 2 spaces

noppanit committed:
Convert tab to spaces again with search and replace

noppanit committed:
Convert tab to spaces again with search and replace

noppanit committed:
Convert tab to spaces again with search and replace

noppanit committed:
Add gulp-rename npm to package.json

noppanit committed:
Mistyped character in gulp build

noppanit commented:
The build is failed because 

```
Error: Cannot find module 'xtend'
```

on module `through2` Have you seen this before?

crookedneighbor commented:
No. Will look at it this weekend.

crookedneighbor commented on tasks/gulp-clean.js:
```diff
@@ -0,0 +1,7 @@
+import gulp from 'gulp';
+
+let rimraf = require('rimraf');
```
This should be `import rimraf from 'rimraf';`

noppanit commented:
Related issue with Travis Failed on CoffeeScript https://github.com/github/generator-hubot/pull/6

noppanit committed:
Use ES6 style to import rimraf

crookedneighbor commented on tasks/gulp-build.js:
```diff
@@ -20,3 +54,5 @@ gulp.task('build:dev:watch', ['build:dev'], () => {
 gulp.task('build:prod', ['babel:common', 'prepare:staticNewStuff'], (done) => {
   gulp.start('grunt-build:prod', done);
 });
+
```
These two extra new lines are not necessary.

noppanit committed:
Finish loadManifestFiles task. It's been migrated to Gulp

noppanit committed:
Add task hashres

noppanit committed:
Merge branch 'develop' of https://github.com/HabitRPG/habitrpg into remove_grunt

noppanit commented:
I have two tasked that I have converted as I see that there's some test tasks and karma in https://github.com/HabitRPG/habitrpg/blob/develop/tasks/gulp-tests.js already. Not sure if the tasks in Gruntfile.js still being used.

noppanit committed:
Rolled back to coffee script 1.6.x instead of 1.7+ because of the backward compatibility

crookedneighbor commented on tasks/gulp-build.js:
```diff
@@ -18,5 +34,95 @@ gulp.task('build:dev:watch', ['build:dev'], () => {
 });
+
+
+function doBrowserify() {
```
Why did you make these normal functions instead of gulp tasks?

noppanit replied:
At first I thought if would be easier, but thinking about it. It's probable easier to read if I make it a gulp task. I'll refactor it. And the style here is one file per gulp task?

crookedneighbor replied:
For this case, yes, I think that'd be best.

noppanit committed:
Separate gulp tasks into files
- gulp-browserify
- gulp-copy
- gulp-cssmin
- gulp-hashres
- gulp-stylus
- gulp-uglify

Refactored to reduce DRY

noppanit committed:
Merged with upstream

noppanit commented:
I've separated all the tasks to smaller ones. I'm not sure why the build is broken.

noppanit committed:
Migrate build:test to gulp instead

noppanit committed:
Try to build in Travis with just gulp

noppanit committed:
Revert "Try to build in Travis with just gulp"

This reverts commit 10a43c600da9a4c97ca11b3f88e9aa0986d6013b.

noppanit committed:
Delete unused grunt packages

noppanit committed:
Merged with upstream

noppanit commented:
@crookedneighbor you mind code review this PR?

crookedneighbor commented on tasks/gulp-uglify.js:
```diff
@@ -0,0 +1,38 @@
+
+import {basename, dirname } from "./taskHelper.js";
+
+let rename = require("gulp-rename");
```
Please use the import syntax for all of these requires.

crookedneighbor commented on tasks/gulp-tests.js:
```diff
@@ -55,13 +55,21 @@ gulp.task('test:prepare:server', ['test:prepare:mongo'], () => {
 
 gulp.task('test:prepare:build', (cb) => {
-  exec(testBin('grunt build:test'), cb);
+  exec(testBin('gulp test:prepare:translations && gulp build:dev'), cb);
```
Why this instead of:

``` js
gulp.task('test:prepare:build', ['test:prepare:translations', 'build:dev']);
```

noppanit replied:
I tried that and somehow the build broke. I'll try that again.

crookedneighbor commented:
Something is wrong with the css compiling:

![screen shot 2015-10-30 at 1 12 44 pm](https://cloud.githubusercontent.com/assets/2916945/10854160/fbc626ae-7f07-11e5-9b93-b7a544757f26.png)

crookedneighbor commented:
Any progress?

noppanit commented:
Sorry was out of town. Should be done today or tmr. :-)

noppanit committed:
Changed shared css and add callbed to gulp-build so the task can finish

noppanit committed:
Merged with upstream

noppanit committed:
Change import style

noppanit committed:
Merge branch 'develop' of https://github.com/HabitRPG/habitrpg into remove_grunt

noppanit committed:
Change import style from gulp-uglify

noppanit committed:
Change import style from gulp-uglify

noppanit committed:
Try using gulp task instead of exec

noppanit committed:
Typo for lodash

noppanit committed:
Typo for merge-stream

noppanit committed:
fixed the path of i18n.js file

noppanit committed:
Revert to call the callback

noppanit committed:
Trigger

noppanit committed:
Build with exec testBin

noppanit committed:
Revert "Build with exec testBin"

This reverts commit 85e3913a0020071693516ad61114a72c828d72f0.

noppanit committed:
Call callback to finish the task

noppanit committed:
Try without callback

noppanit commented:
@crookedneighbor apologies I'm pretty new at this project. I'm not sure what's causing the wrong css.  Can you help checking? I might be that I might have missed something.

noppanit commented:
After I compare the new css, the positions for sprites are off

noppanit committed:
Merge branch 'develop' of https://github.com/HabitRPG/habitrpg into remove_grunt

noppanit committed:
Revert "Try without callback"

This reverts commit 1668f225b53c41faa99cf60df0340bac2dcad296.

noppanit committed:
Run task with exec instead. It seems gulp tasks have to be rearranged to make them async or sync

gauthamchandra commented:
@noppanit, @crookedneighbor I am not too familiar with the codebase but I am with all the technologies and the stack involved. Tell me if you need help with this.

gauthamchandra commented:
There needs to be a lot of cleanup with all the npm modules we use. Node is up to v5.0 and many of the modules we use requires `v0.10.x` and give a ton of deprecated warnings.

Once this PR is merged, I will file a new issue and a matching PR to clean up our dependencies and do a full check to make sure our code is properly linted and follows best practices

paglias commented:
@gauthamchandra thanks for help!

A few things to keep in mind, though:
1. We're going to upgrade to Node v4.2, which is an LTS, not to Node 5. We might review it later but for the time being we'll switch to Node 4.2
2. We're in the process of upgrading most dependencies, switching to ES6 with ESLint for code style and re-writing the API completely in a separate PR https://github.com/HabitRPG/habitrpg/pull/6144
3. We'll be happy to accept any PR that upgrades dependencies to the newest version as long as it doesn't break compatibility with API v2 and hasn't already been addressed in https://github.com/HabitRPG/habitrpg/pull/6144

Let me know if you have any question

noppanit commented:
@gauthamchandra I'm going to merge the conflict so please feel free to jump in this PR. I have two problems
1. https://github.com/HabitRPG/habitrpg/pull/6112/files#diff-565c7aaf8560ac8a12c341fa07bed43aR60 I can't seem to use 

`gulp.task('test:prepare:build', ['test:prepare:translations', 'build:dev']);` in Travis. The process wouldn't exit I suspect it's because some of the tasks hanging and wouldn't exit and Travis killed the build.
1. I can't seem to figure out the discrepancy of the `cssmin` between `gulp` and `grunt`. The generated css seems to be broken in `gulp`

Thanks for your help

noppanit committed:
Merged with upstream

gauthamchandra commented:
@paglias wow that is a massive set of changes in the new PR. I will forgo the changes and look into that PR instead. :+1:

noppanit commented:
@gauthamchandra Sorry I was on the road so didn't have a chance to catch up. Were you able to take a look at the PR?

noppanit committed:
Merged with master

gauthamchandra commented:
@noppanit I got too busy over the weekend and wasn't able to look at it. I will try looking over it in the next 1-2 days.

noppanit commented:
@crookedneighbor I think this PR is not going to work anymore as it's far behind the master. Is it ok if I create a new PR?

crookedneighbor commented:
Let's do smaller PRs extracting one grunt task at a time.

noppanit commented:
Will do that.
